### PR TITLE
Revise expression parsing

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Recent forways into fixing expression parsing issues have not been fun.  The core expression precedence parser is written in a very convoluted and hard to follow way.  In particular, it tends to use a lot of local state variables for different purposes, with lack of clarity about the state tehy represent.  Portions of that operation also mutate the state of the parser (sometimes consumign only parts of hte input, with the later locations required to figure that out and do the right thing).    There are often relationships between variables which are hard to track and reason about.  

This is a lot of complexity for what is really just an iterative looping algorithm for precedence parsing.  This PR keeps the iterative nature of parsing here (we do *not* want recursive decent for expressions as we absolutely will blow the stack), while greatly simplifying the individual steps that parsing needs to perform.  

Comments inline as to the changes and reasons for them.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -10937,9 +10937,6 @@ private ExpressionSyntax ParseSubExpression(Precedence precedence)
 
         private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
         {
-            ExpressionSyntax leftOperand;
-            Precedence newPrecedence = 0;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10941]: an example of very confusing local state.  inner blocks would set this, to use only within that block.  But the value was never read afterwards.  Removed entirely as having multiple precedence values to have to understand in a precedence parser makes things much harder to reason about :)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -10949,99 +10946,143 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
-            if (IsExpectedPrefixUnaryOperator(tk))
-            {
-                var opKind = SyntaxFacts.GetPrefixUnaryExpression(tk);
-                newPrecedence = GetPrecedence(opKind);
-                var opToken = this.EatToken();
-                var operand = this.ParseSubExpression(newPrecedence);
-                leftOperand = _syntaxFactory.PrefixUnaryExpression(opKind, opToken, operand);
-            }
-            else if (tk == SyntaxKind.DotDotToken)
+            return ParseExpressionContinued(parseUnaryOrPrimaryExpression(precedence), precedence);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10951]: extracted out common patters to make things much clearer.  The core algorithm is now simple to reason about.  We first parse out a unary/primary expression.  Then, we parse out what can continue that at the current precedence we're at.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -10949,99 +10946,143 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
-                var opToken = this.EatToken();
-                newPrecedence = GetPrecedence(SyntaxKind.RangeExpression);
+                // Parse a left operand -- possibly preceded by a unary operator.
+                if (IsExpectedPrefixUnaryOperator(tk))
+                {
+                    var opKind = SyntaxFacts.GetPrefixUnaryExpression(tk);
+                    return _syntaxFactory.PrefixUnaryExpression(
+                        opKind,
+                        this.EatToken(),
+                        this.ParseSubExpression(GetPrecedence(opKind)));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10963]: moved to a model where instead of assinging to a variable (and then having to see what happens to it), we have a local function which just returns the value in question.  Making it much easier to see that this is just responsible for parsing out that construct and *nothing else*.

also, moved to the much clearer parsing pattern we try to use int eh rest of hte parser where the production `A -> B C D` is parsed as `return A(ParseB(), ParseC(), ParseD())`.  this lowers the number of locals to keep track of and ensures not accidentally using them for an innapropriate purpose.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -10949,99 +10946,144 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
-                if (CanStartExpression())
+                if (tk == SyntaxKind.DotDotToken)
                 {
-                    rightOperand = this.ParseSubExpression(newPrecedence);
+                    return _syntaxFactory.RangeExpression(
+                        leftOperand: null,
+                        this.EatToken(),
+                        CanStartExpression()
+                            ? this.ParseSubExpression(Precedence.Range)
+                            : null);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10973]: example of simplifying.  instead of computing the nested precedence dynamically (even though it is fixed), and assigning to an outer varaible (With lack of clarity who is affected by that), we simply parse the subexpression with the correct precedence for this construct.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10973]: note: i'm fairly certain we don't need this.  a range expression is not a primary or unary expression.  this should get picked up in teh 'continur parsing' routine.  will try that out.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10973]: by naming the functions appropriately, this becomes clearer.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -10949,99 +10946,144 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
+                        ? result
+                        : this.AddError(result, ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
+                }
+
+                if (this.IsPossibleDeconstructionLeft(precedence))
+                    return ParseDeclarationExpression(ParseTypeMode.Normal, isScoped: false);
+
                 // Not a unary operator - get a primary expression.
-                leftOperand = this.ParseTerm(precedence);
+                return this.ParsePrimaryExpression(precedence);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11009]: base (primary) case now falls out at the bottom once unary cases

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 6]
```diff
@@ -10949,99 +10946,144 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
+            //     | null_coalescing_expression '?' expression ':' expression
+            //     ;
+            //
+            // 1. Only take the conditional part of the expression if we're at or below its precedence.
+            // 2. When parsing the branches of the expression, parse at the highest precedence again ('expression').
+            //    This allows for things like assignments/lambdas in the branches of the conditional.
+            if (this.CurrentToken.Kind == SyntaxKind.QuestionToken && precedence <= Precedence.Conditional)
+                return consumeConditionalExpression(expandedExpression);
+
+            return expandedExpression;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11041]: the entirety of hte ParseExpressionCOntineud method is now much simpler.  You start with the initial unary/primary expression passed in, and your current precedence.  It then proceeds to try to expand the expression as long as it can, finally bottoming out with `? :` parsing if allowed.  Much simpler to understand, esp. in comparison to our actual spec.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -10949,99 +10946,144 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
+
+            bool tryExpandExpression(ref ExpressionSyntax leftOperand, Precedence precedence)
             {
-                // We either have a binary or assignment operator here, or we're finished.
+                // Look for operators that can follow what we've seen so far, and which are acceptable at this
+                // precedence level.  Examples include binary operator, assignment operators, range operators `..`, as
+                // well as `switch` and `with` clauses.
                 var tk = this.CurrentToken.ContextualKind;
 
-                bool isAssignmentOperator = false;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11030]: example of ambient state that was hard to keep track of.  we can compute this on demand trivially in the one place later that needs this.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -10949,99 +10946,144 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
 
                 // If the set of expression continuations is updated here, please review ParseStatementAttributeDeclarations
                 // to see if it may need a similar look-ahead check to determine if something is a collection expression versus
                 // an attribute.
 
-                if (IsExpectedBinaryOperator(tk))
+                // check for >>, >>=, >>> or >>>=
+                if (tk == SyntaxKind.GreaterThanToken
+                    && this.PeekToken(1) is { Kind: SyntaxKind.GreaterThanToken or SyntaxKind.GreaterThanEqualsToken } token1
+                    && NoTriviaBetween(this.CurrentToken, token1)) // check to see if they really are adjacent
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11059]: instead of having `>` be handled by 'IsBinaryOperator' and then undoing its logic.  we just check for the fun cases up front, and then only handle it as a real > if it doesn't match that.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 9]
```diff
@@ -11057,67 +11099,28 @@ private ExpressionSyntax ParseExpressionContinued(ExpressionSyntax leftOperand,
...
-                    break;
+                    // Something that doesn't expand the current expression we're looking at.  Bail out and see if we
+                    // can end with a conditional expression.
+                    return false;
                 }
 
                 var newPrecedence = GetPrecedence(opKind);
 
-                // check for >>, >>=, >>> or >>>=
-                int tokensToCombine = 1;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11066]: also a difficult value to reason about, with complex semantics about hwo it gets set.  no need for it at all.  We have opKind.  So we know precisely what tokens we have and if we have to merge them.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 10]
```diff
@@ -11349,7 +11352,7 @@ private ExpressionSyntax ParseIsExpression(ExpressionSyntax leftOperand, SyntaxT
             };
         }
 
-        private ExpressionSyntax ParseTerm(Precedence precedence)
+        private ExpressionSyntax ParsePrimaryExpression(Precedence precedence)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11355]: clarifying name.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 11]
```diff
@@ -11057,67 +11102,28 @@ private ExpressionSyntax ParseExpressionContinued(ExpressionSyntax leftOperand,
                 }
                 else
                 {
-                    break;
+                    // Something that doesn't expand the current expression we're looking at.  Bail out and see if we
+                    // can end with a conditional expression.
+                    return false;
                 }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11108]: note: we could consider extracing this into a tryGetOperatorKind helper, to get similar benefits on single-exit logic.  not sure if it is needed thoughl

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 12]
```diff
@@ -11349,165 +11355,263 @@ private ExpressionSyntax ParseIsExpression(ExpressionSyntax leftOperand, SyntaxT
...
+        private ExpressionSyntax ParsePrimaryExpression(Precedence precedence)
         {
-            var tk = this.CurrentToken.Kind;
-            switch (tk)
+            // Primary expressions:
+            // x.y, f(x), a[i], x?.y, x?[y], x++, x--, x!, new, typeof, checked, unchecked, default, nameof, delegate, sizeof, stackalloc, x->y
+            //
+            // Note that postfix operators (like ++) are still primary expressions, even though their prefix equivalents (`++x`) are unary.
+
+            return parsePostFixExpression(parsePrimaryExpressionWithoutPostfix(precedence));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11365]: Inlined these methods since no one else shoudl call them.  now it's clearer that parsing a primary is just prsing the initial part, and the the postfix portion.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 13]
```diff
@@ -11349,165 +11355,263 @@ private ExpressionSyntax ParseIsExpression(ExpressionSyntax leftOperand, SyntaxT
...
+                                expr = this.AddError(expr, ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
+                            }
+
+                            return expr;
                         }
+                }
+            }
 
-                        return expr;
+            ExpressionSyntax parsePostFixExpression(ExpressionSyntax expr)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11526]: inlined into ParsePrimaryExpression.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 14]
```diff
@@ -11349,165 +11355,263 @@ private ExpressionSyntax ParseIsExpression(ExpressionSyntax leftOperand, SyntaxT
...
+                            }
+                            else
+                            {
+                                expr = this.AddError(expr, ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
+                            }
+
+                            return expr;
                         }
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11525]: bad github diff.  ParseBaseExpression, IsPossibleDeconstructionLeft and IsPossibleAnonymousMethodExpression are entirely unchanged.  they just moved below the code that was now inlined into this method.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler This is ready for review. Changes that really helped make it easier to reason about expression parsing while i was fixing the incremental parsing bug with `..` expressions.  I have extracted out so we can independently take the incremental parsing fix, while also getting the benefits of clearer and easier to maintain parsing.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10976]: > this should get picked up in teh 'continur parsing' routine. will try that out.

Interested in finding out the result of this.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 15]
```diff
@@ -11180,117 +11185,118 @@ private ExpressionSyntax ParseExpressionContinued(ExpressionSyntax leftOperand,
...
+            {
+                // Combine tokens into a single token if needed
+                if (operatorKind is SyntaxKind.RightShiftExpression or SyntaxKind.RightShiftAssignmentExpression)
                 {
-                    // Operator ".." here can either be a binary or a postfix unary operator:
-                    Debug.Assert(opKind == SyntaxKind.RangeExpression);
+                    // >>  or >>=
+                    return SyntaxFactory.Token(
+                        this.EatToken().GetLeadingTrivia(),
+                        operatorKind == SyntaxKind.RightShiftExpression ? SyntaxKind.GreaterThanGreaterThanToken : SyntaxKind.GreaterThanGreaterThanEqualsToken,
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11210]: It feels like we are translating the operator-token-kind into an expression-kind in a previous step, and then here doing the same translation in the other direction. Would it be simpler to just pass along the operator-token-kind? It's possible I am missing something.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 11 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11108]: What I am finding as a reviewer is that the repeated "if-return-if-return-if-return"s are much easier to understand at a glance than the repeated "if-else-if-else-if-else-if". Both here and below, where the `leftOperand` is assigned.

This is in part because the mental compiler needs to locate where we are going to jump to after exiting a large if-else block, scroll to it, decide whether any additional work is being done, decide whether we are in a consistent state for doing that work, etc.

Also, I don't think there's anyway to "fix" this, but, I am finding that scrolling between corresponding places in the old and new versions, to find where logic has moved to, is quite challenging, to the point that I might suggest to the second reviewer to just skim thru the deleted lines, then review the new change practically as new code.

**Review by RikkiGibson — APPROVED** [review]
Change LGTM but it would be nice to see this code pushed even farther toward making it easy to understand.

**Comment by CyrusNajmabadi** [comment]
> Interested in finding out the result of this.

I was slightly incorrect.  `expr ..` is picked up in the 'contuation part'.  But we still need something to handle the base case of `..` starting an expr.  

Note: the logic is still a *little* incorrect.  Specifically, we should only take the 'range' here if precedence allows it.  That said, no one calls into ParseSubExpressionCore with a precedence tighter than 'Range', so it never matters in practice.  I'm considering cleaning this up further in a future pass, but would like to leave this as is for now.  I have commented to the code explaining things.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 15 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11210]: Let me see about that!

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This did improve things.  Thanks!

**Comment by CyrusNajmabadi** [comment]
> What I am finding as a reviewer is that the repeated "if-return-if-return-if-return"s are much easier to understand at a glance

Yup.  I changed the leftOperand code to follow that pattern too.  much nicer :)

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  for another pair of eyes.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 16]
```diff
@@ -10937,9 +10937,6 @@ private ExpressionSyntax ParseSubExpression(Precedence precedence)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10937]: i recommend viewing with whitespace off as some sibling methods became local funcitons.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10976]: answered in main thread.  and doc'ed here.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 17]
```diff
@@ -11130,44 +11085,42 @@ private ExpressionSyntax ParseExpressionContinued(ExpressionSyntax leftOperand,
...
-                    Debug.Assert(opToken2.Kind == SyntaxKind.GreaterThanToken);
-                    opToken2 = this.EatToken();
-                    var kind = opToken2.Kind == SyntaxKind.GreaterThanToken ? SyntaxKind.GreaterThanGreaterThanGreaterThanToken : SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken;
-                    opToken = SyntaxFactory.Token(opToken.GetLeadingTrivia(), kind, opToken2.GetTrailingTrivia());
-                }
-                else if (tokensToCombine != 1)
-                {
-                    throw ExceptionUtilities.UnexpectedValue(tokensToCombine);
+                    return _syntaxFactory.BinaryExpression(
+                        operatorExpressionKind, leftOperand, operatorToken, this.ParseType(ParseTypeMode.AsExpression));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11097]: in a similar vein, these also just became simple returns, instead of assignments to track.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 18]
```diff
@@ -11116,12 +11063,12 @@ bool tryExpandExpression(ref ExpressionSyntax leftOperand, Precedence precedence
                     return false;
 
                 // Same precedence, but not right-associative -- deal with this "later"
-                if ((newPrecedence == precedence) && !IsRightAssociative(opKind))
+                if ((newPrecedence == precedence) && !IsRightAssociative(operatorExpressionKind))
                     return false;
 
-                // We'll "take" this operator, as precedence is tentatively OK.
+                // No consume the operator (including consuming multiple tokens in the case of merged operator tokens)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11071]:
```suggestion
                // Now consume the operator (including consuming multiple tokens in the case of merged operator tokens)
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 19]
```diff
@@ -10949,175 +10946,133 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
-
-            return ParseExpressionContinued(leftOperand, precedence);
         }
 
-        private ExpressionSyntax ParseExpressionContinued(ExpressionSyntax leftOperand, Precedence precedence)
+#nullable enable
+
+        /// <summary>
+        /// Takes in an initial unary expression or primary expression, and then consumes what follows as long as as
+        /// it's precedence is either lower than the <paramref name="precedence"/> we're parsing currently (or equal, to
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11020]: >currently (or equal, to [](http://example.com/codeflow?start=97&length=23)

Minor: Perhaps drop the open paren since there is no close paren and move the comma.
> currently, or equal to

Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11020]: "as long as its"

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Done.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 20]
```diff
@@ -10949,175 +10946,133 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
...
-                            tokensToCombine = 2;
-                        }
-                    }
-                    else
-                    {
-                        opKind = SyntaxFacts.GetAssignmentExpression(SyntaxKind.GreaterThanGreaterThanEqualsToken);
-                        isAssignmentOperator = true;
-                        tokensToCombine = 2;
-                    }
+                var (operatorTokenKind, operatorExpressionKind) = getOperatorTokenAndExpressionKind();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11054]: >operatorTokenKind, operatorExpressionKind [](http://example.com/codeflow?start=21&length=41)

Consider calling these `opToken` and `opKind` to reduce the differences.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11054]: i intentionally changed as the old names were more confising.  opToken was not a token, but was a kind.  opKind was not the kind of the op token itself, but of the op expression.  

The new names are much clearer (both are kinds) and indicate which domain they're describing (the token domain or the expression domain).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 21]
```diff
@@ -11180,117 +11133,178 @@ private ExpressionSyntax ParseExpressionContinued(ExpressionSyntax leftOperand,
...
+                    return SyntaxFactory.Token(
+                        token1.GetLeadingTrivia(),
+                        operatorTokenKind,
+                        token3.GetTrailingTrivia());
                 }
                 else
                 {
-                    Debug.Assert(IsExpectedBinaryOperator(tk));
-                    leftOperand = _syntaxFactory.BinaryExpression(opKind, leftOperand, opToken, this.ParseSubExpression(newPrecedence));
+                    // Normal operator.  Eat as a single token (converting cases contextual words 'with' to a keyword)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11237]: >converting cases contextual words [](http://example.com/codeflow?start=64&length=33)

Perhaps "converting contextual keyword".

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11237]:
```suggestion
                    // Normal operator.  Eat as a single token, converting contextual words cases (like 'with') to a keyword.
```

**Review by cston — APPROVED** [review]
